---
title: "Wrangling shotgun metagenomic data for analysis with _phyloseq_"
author: "Miljenka Vuko"
date: "12/9/2019"
output:
  html_document:
    fig_height: 4
    fig_width: 5
    fig_align: center
    highlight: breezedark
    theme: darkly
    toc: yes
    toc_depth: 3
    toc_float:
      collapsed: no
      smooth_scroll: no
      keep_md: yes
    includes:
      after_body: footer.html
---


# 1. Introduction

**_phyloseq_** (McMurdie & Holmes, 2013) is a package created for the analysis of high-throughput microbiome data. Specifically, it was designed to import, analyze, and graphically display phylogenetic sequencing data clustered into Operational Taxonomic Units (OTUs), or Amplicon Sequence Variants (ASVs), depending on the raw sequence analysis. _phyloseq_ uses a specialized system of S4 classes, A.K.A phyloseq objects, to store all necessary data and metadata into a single object. This makes working with complex microbiome datasets much easier. Read more [here](https://joey711.github.io/phyloseq/). 


As opposed to amplicon sequencing of the 16S rRNA gene used for phylogenetic studies, taxonomic annotation of reads after shotgun metagenomic sequencing does not result in OTUs or ASVs. However, datasets do have a similar structure, i.e. taxonomy, read counts, and metadata, making it possible to use tools such as _phyloseq_ for an easier analysis. 


* Therefore, this walkthrough shows how to:
  + wrangle non-OTU sequencing data using **_tidyverse_** (Wickham _et al._, 2019) into data that _phyloseq_ can use as an input,
  + initially inspect data with ordination plots and stacked barplots.

  

Data used for this walkthrough (Table 1) originates from shotgun metagenomic sequencing reads that were taxonomically classified using Kaiju (Menzel _et. al._, 2016) and the NCBI-nr database.[^1]


[^1]: When using an approach of this kind to analyze metagenomic data, it might be worth to extract 16S reads and analyze them for comparison (however, the number of reads assigned to 16S in metagenomic datasets is usually quite low). 


**Data structure**



After classification, the taxonomy table looks like this:

![](otu_taxa.table.png)

Table 1


The first column (SampleID) is taxonomy, which is delimited by semicolons, and other columns are sample names.


The matching metadata table looks like this:

![](metadata.png)

Table 2

* _phyloseq_ requires  three tables:
  + an OTU table
  + a taxonomy table 
  + a metadata table.



# 2. Data input and wrangling

First, we will load the libraries and then input the tables.

As Table 1 consists of both taxonomy and the number of assigned reads, we will first input the whole table, and then split it up so that the first column becomes a taxonomy table, and the other columns become an OTU table.
Then we will input the metadata table, and create a _phyloseq_ object that unifies all three.

```{r, warning= FALSE, message= FALSE, results= "hide"}
library(tidyverse)
library(phyloseq)
```



### 2.1. OTU table

OTUs have to be row names and therefore unique, which is the crucial feature. In Table 1, the row names are not OTU names, but they are all unique, so we will use this first column of Table 1 as the row names for the OTU table.

```{r, collapse=TRUE}
otu<- read.delim("otutable.species.txt", sep = "\t", header = TRUE)

#always check the structure of the object to make sure everything is as you expect
str(otu)

#OTU ID's have to be the rownames
rownames(otu)<- otu$SampleID

#check if the number of reads per sample is correct 
# (e.g. sample RA0_A has 1364214 bacterial reads, so make sure that is correct)
count <- colSums(otu[, c(2:ncol(otu))])

count
```

### 2.2. Taxonomy table

The taxonomy table is created from the SampleID column of Table 1. This column consists of taxonomy levels separated by semicolons, which we will split into different columns.

```{r, collapse=TRUE}
taxa<- otu %>% 
  select(SampleID) %>% 
  separate(SampleID, c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species"),
           "; ")  

 
#This warning appears because the last row are Unassigned reads, and they do not have taxa levels.

str(taxa) 

#the output is a data frame of characters, and we need taxa to be recognized as factors
taxa<- taxa %>% 
  mutate_if(is.character, as.factor)

str(taxa)


#add the first column from the OTU table so that OTU and taxa tables match
taxa<- cbind(otu$SampleID, taxa)

#rename the first column
colnames(taxa)[1]<- "SampleID"
str(taxa)

#OTU ID's have to be the rownames, same as in the OTU table
rownames(taxa)<- taxa$SampleID

#now you can delete the first columns of both OTU and taxa tables
# because they are now the rownames, and therefore redundant in the first column
otu<- otu %>% 
  select(-SampleID)

taxa<- taxa %>% 
  select(-SampleID)
```

### 2.3. Metadata table

The metadata table contains a description of sample names (Table 2). This can include everything that describes a sample; sample origin, type, sampling time etc., depending on the experimental design.  
Samples in this walkthrough are soil samples taken from fields of agricultural reclamation sites of different age, e.g. RA1_A = one year old reclamation field, plot A.
  
```{r, collapse=TRUE}
meta<- read.delim("metadata.txt", sep = "\t", header = TRUE)
str(meta)

#make sample names row names 
rownames(meta)<- meta$SampleID

#and delete the first column because it is now redundant
meta<- meta %>% 
        select(-SampleID)

#make proper levels of the factor so that they are in the wanted order in figures
meta$Age<- factor(meta$Age, levels = c("RA0", "RA1", "RA3"))
meta$Plot<- factor(meta$Plot, levels = c("a", "b", "c"))
```


# 3. _phyloseq_

OTU and taxonomy tables have to be matrices, and then all three tables have to be transformed to phyloseq objects before unifying them into one object.

```{r, collapse=TRUE}
otu_mat<- as.matrix(otu)
tax_mat<- as.matrix(taxa)

#transform data to phyloseq objects
phylo_OTU<- otu_table(otu_mat, taxa_are_rows = TRUE)
phylo_TAX<- tax_table(tax_mat)
phylo_samples<- sample_data(meta)

#and put them in one object
phylo_object<- phyloseq(phylo_OTU, phylo_TAX, phylo_samples)

#check if everything looks good
sample_sums(phylo_object)       #should sum up the number of all reads per sample

sample_names(phylo_object)      #sample names

rank_names(phylo_object)        #taxa levels  

sample_variables(phylo_object)  #factors

otu_table(phylo_object)[1:3, 1:2]

taxa_names(phylo_object)[1:5]
```


## 3.1. Relativization

Subsequent analyses are based on relative data. Since here we have only the reads that were assigned to Bacteria, we will relativize them to the total number of bacterial reads per sample and show them as percentages.

```{r fig.height=5, fig.width=5, collapse=TRUE}
phylo_rel<- transform_sample_counts(phylo_object, function(x) x*100 / sum(x))

#check if everything looks good
sample_sums(phylo_rel)          #should sum up to 100% per sample

sample_names(phylo_rel)         #sample names

rank_names(phylo_rel)           #taxa levels  

sample_variables(phylo_rel)     #factors

otu_table(phylo_rel)[1:3, 1:2]

taxa_names(phylo_rel)[1:5]

#to start getting the feel of the data, let´s check what had the highest abundance

max(phylo_rel@otu_table)        #the highest relative abundance

rownames(phylo_rel@otu_table)[which.max(apply(phylo_rel@otu_table,MARGIN=1,max))] #row name
# of the highest relative abundance

#the NAs appear when reads were not assigned down to the respective taxonomy level.  

phylo_rel@otu_table['Bacteria; Acidobacteria; NA; NA; NA; NA; NA'] #extract that whole row
```

## 3.2. Visual inspection of the data

### 3.2.1. Ordination

To get the first visual overview of the data, we will plot an ordination. _phyloseq_ has built-in functions (incorporated from the _vegan_ package (Oksanen _et al._, 2019)), which calculate different ordinations based on different distance matrices.
Here, we will make a non-metric multidimensional scaling (NMDS) and a PCoA ordination based on the Bray-Curtis dissimilarity and then plot them.

```{r, collapse=TRUE, fig.align="center"}
#NMDS plot
phylo_rel_nmds<- ordinate(phylo_rel, method = "NMDS", distance = "bray")


#we can check what the stressplot (Shepard plot) looks like
vegan::stressplot(phylo_rel_nmds)
#R2 is 1, but the plot looks strange and the warning after the NMDS calculation suggest that NMDS
# might not be the best choice for this data.

#NMDS plot
plot_ordination(phylo_rel, phylo_rel_nmds,
                color = "Age", shape = "Plot") +
  theme_bw()
#PCoA
phylo_rel_pcoa<- ordinate(phylo_rel, method = "PCoA", distance = "bray")

#PCoA plot
plot_ordination(phylo_rel, phylo_rel_pcoa,
                color = "Age", shape = "Plot") +
  theme_bw()
```


### 3.2.2. Stacked barplots

For stacked barplots, we will work with means of replicates using the `merge_samples` function. The default of this function is `merge_samples(x, group, fun= mean)`. 

**However**, in further description of the arguments, specifically `fun`, it says:

>Default is mean. Note that this is (currently) ignored for the otu_table, where the equivalent function is sum, but evaluated via rowsum for efficiency.

Therefore, in order to really get the average, we have to divide the result we get with the default function by the number of replicates.

```{r, collapse=TRUE}
phylo_rel_mean <- merge_samples(phylo_rel, "Age")

sample_sums(phylo_rel_mean) #sums up to 300!

#now divide all relative OTU abundances by the number of replicates
phylo_rel_mean<- transform_sample_counts(phylo_rel_mean, function(x) x / 3)

sample_sums(phylo_rel_mean)
```

Stacked barplots can show taxonomic levels of choice. We will create a plot that shows the phylum level.
First, we will agglomerate data to the phylum level using the `tax_glom` function, and then plot the agglomerated data.

Be careful when using the `tax_glom` function because the default behaviour is to remove NAs (those which are unassigned on the desired level), or agglomerate all of them together. This also includes empty cells or any strange/bad cells. Read more about this on `?tax_glom` before use.
Always check this using the `sample_sums` function, and by plotting the resulting object. With stacked barplots, the NA stack will be large if there are a lot of NAs. 


You can fix this by manually changing the cells. For example, if you are doing an analysis of highly abundant bacterial families and you end up with a lot of NAs, you can check the taxonomy table and see rows that look something like this:

> 'Bacteria; Chloroflexi; NA; NA; NA; NA; Chloroflexi bacterium OLB14'

You can manually change this row to replace the NA with a value in the family column:

> physeq_top_fam@tax_table['Bacteria; Chloroflexi; NA; NA; NA; NA; Chloroflexi bacterium OLB14','Family']<- "uncultured Chloroflexi"


By doing this, the NA stack in your stacked barplot will shrink, and a new "uncultured Chloroflexi" stack will appear. You can do this for as many rows as you want.



```{r, collapse=TRUE, fig.align="center"}
#check taxonomy level names
rank_names(phylo_rel_mean)

# agglomeration on the phylum level
phylo_rel_rel_mean_phylum<- tax_glom(phylo_rel_mean, taxrank = "Phylum")

#check the number of taxa in the whole phyloseq object
phylo_rel_mean 

#check to how many phyla were these assigned to
phylo_rel_rel_mean_phylum 

#check sample sums to make sure nothing was deleted due to NAs (should sum up 100%)
sample_sums(phylo_rel_rel_mean_phylum)

plot_bar(phylo_rel_rel_mean_phylum) +
  theme_bw()
#if we try to color by Phylum to see their names, 
# we can see that 135 are too many to see in a stacked barplot
plot_bar(phylo_rel_rel_mean_phylum, fill = "Phylum")

```

Since 135 phyla are too many to plot in a stacked barplot, we will filter the low abundant ones and put them into one category. To do this, we will use the power of _tidyverse_ again.
First, we will create a normal data frame out of the phyloseq object and then add another column where all taxa with abundance lower than 1% will be renamed to "< 1%".

```{r, collapse=TRUE, fig.align="center"}
#transform phyloseq object to a data frame
phylo_rel_rel_mean_phylumDF<- psmelt(phylo_rel_rel_mean_phylum)

str(phylo_rel_rel_mean_phylumDF)

#make the phyla characters, not factors
phylo_rel_rel_mean_phylumDF$Phylum<- as.character(phylo_rel_rel_mean_phylumDF$Phylum)

#add new column with renamed low abundant taxa
phylo_rel_rel_mean_phylumDF<- phylo_rel_rel_mean_phylumDF %>% 
  mutate(Phylum2 = replace(Phylum, Abundance < 1, "< 1%"))

#check all phyla names
unique(phylo_rel_rel_mean_phylumDF$Phylum2)

#there are some reads that were assigned only to the domain level, 
# i.e. NA on the phylum level, so we will rename them
phylo_rel_rel_mean_phylumDF<- phylo_rel_rel_mean_phylumDF %>% 
  mutate(Phylum2 = replace(Phylum2, Phylum2 == "NA", "unassigned Bacteria"))

#reorder the phyla so that they are stacked according to abundance
phylo_rel_rel_mean_phylumDF$Phylum2<- reorder(phylo_rel_rel_mean_phylumDF$Phylum2,
                                              phylo_rel_rel_mean_phylumDF$Abundance)

#check how many unique phyla are there to find discrete colors for them
unique(phylo_rel_rel_mean_phylumDF$Phylum2)

ggplot(phylo_rel_rel_mean_phylumDF, aes(Sample, Abundance, fill=Phylum2)) +
  geom_bar(stat = "identity") +
  scale_fill_manual(values = c("#bd6b8f",
                              "#6db543",
                              "#7661cd",
                              "#c0b047",
                              "#c35abc",
                              "#60bf8b",
                              "#d13f73",
                              "#3e8149",
                              "#ca5340",
                              "#45b0cf",
                              "#cc8444",
                              "#7882c9",
                              "#7a7732")) +
  labs(y= "Relative abundance [%]",
       fill= "Phlya") +
  theme_bw()

```

Tip: it can be rather difficult to find acceptable colors when you have many discrete variables. Check out this great open-source tool, which allows you to randomly generate as many colors as you need, _and_ gives them in a nice, easily copy-pasteable form - https://medialab.github.io/iwanthue/



# 4. Conclusion

There are a lot of bioinformatic tools out there that were designed for specific data types. If we learn how to transform and adapt our data, we can use the power of great open-source tools without having to "reinvent the wheel" for every new dataset.

However, one should always use these tools with care. Most of us probably don´t look "under the hood" to check how every function works, which can result in mistakes or unwanted outcomes. To minimize this, one should always strive to be mindful during the analysis, and double-check all the steps to make sure the output is reasonable and correct.


## Acknowledgments


A big thank you to everyone who posts their problems and solutions on the internet, and to the wonderful #rstats online community. Special thanks goes to Dr. Antonios Michas for discussions and ideas, and Johan Sebastián Sáenz for great feedback and suggestions that improved this walkthrough.


## References

McMurdie, P. J., Holmes, S. (2013) pyloseq: An R package for reproducible interactive analysis and graphics of microbiome census data. _PLoS ONE_ 8(4):e61217.

Menzel, P., Ng, K.L., and Krogh, A. (2016) Fast and sensitive taxonomic classification for metagenomics with Kaiju. _Nature Communications_ 7: 11257.

Jari Oksanen, F. Guillaume Blanchet, Michael Friendly, Roeland Kindt, Pierre Legendre, Dan McGlinn, Peter R. Minchin, R. B. O'Hara, Gavin L. Simpson,
Peter Solymos, M. Henry H. Stevens, Eduard Szoecs and Helene Wagner (2019). vegan: Community Ecology Package. R package version 2.5-6.

Wickham et al. (2019) Welcome to the tidyverse. _Journal of Open Source Software_, 4(43), 1686.



## Operation and session info
```{r}
sessionInfo()
```
